specs/experimental/inbox-contract.md

```diff
@@ -0,0 +1,189 @@
+3. Modification of the op-batcher submission process: The op-batcher will be updated to [call `recordFailedTx`](https://github.com/blockchaindevsh/optimism/blob/02e3b7248f1b590a2adf1f81488829760fa2ba03/op-batcher/batcher/driver.go#L537) for failed batch transactions. This modification ensures that the data contained in failed transactions will be resubmitted automatically.
+   1. Most failures will be detected during the [`EstimateGas`](https://github.com/ethereum-optimism/optimism/blob/8f516faf42da416c02355f9981add3137a3db190/op-service/txmgr/txmgr.go#L266) call. However, under certain race conditions, failures may occur after the transaction has been included in a block.
+
+These modifications aim to enhance the security and efficiency of the batch submission and processing pipeline, allowing for more flexible and customizable conditions while maintaining the integrity of the derived state.
+
+
+## Supporting Dynamic Updates to Inbox Address
+
+### SystemConfig
+
+The `SystemConfig` is the source of truth for the address of inbox. It stores information about the inbox address and passes the information to L2 as well. 
+
+
+#### setBatchInbox
+
+A new function `setBatchInbox` is introduced to the `SystemConfig` contract, enabling dynamic updates to the inbox:
+
+```solidity
+/// @notice Updates the batch inbox address. Can only be called by the owner.
+/// @param _batchInbox New batch inbox address.
+function setBatchInbox(address _batchInbox) external onlyOwner {
+    _setBatchInbox(_batchInbox);
+}
+
+/// @notice Updates the batch inbox address.
+/// @param _batchInbox New batch inbox address.
+function _setBatchInbox(address _batchInbox) internal {
+    Storage.setAddress(BATCH_INBOX_SLOT, _batchInbox);
+
+    bytes memory data = abi.encode(_batchInbox);
+    emit ConfigUpdate(VERSION, UpdateType.BATCH_INBOX, data);
+}
+
+```
+
+
+#### initialize
+
+The `SystemConfig` now emits an event when the inbox is initialized, while retaining its existing support for inbox configuration during initialization.
+
+```solidity
+function initialize(
+    address _owner,
+    uint32 _basefeeScalar,
+    uint32 _blobbasefeeScalar,
+    bytes32 _batcherHash,
+    uint64 _gasLimit,
+    address _unsafeBlockSigner,
+    ResourceMetering.ResourceConfig memory _config,
+    address _batchInbox,
+    SystemConfig.Addresses memory _addresses
+)
+    public
+    initializer
+{
+    ...
+    // Storage.setAddress(BATCH_INBOX_SLOT, _batchInbox);
+    // initialize inbox by `_setBatchInbox` so that an event is emitted.
+    _setBatchInbox(_batchInbox);
+    ...
+}
+```
+
+#### UpdateType
+
+A new enum value `BATCH_INBOX` is added to the `UpdateType` enumeration.
+
+```solidity
+enum UpdateType {
+    BATCHER,
+    GAS_CONFIG,
+    GAS_LIMIT,
+    UNSAFE_BLOCK_SIGNER,
+    BATCH_INBOX
+}
+```
+
+### L1Block
```

Comment:
I would suggest following the description of custom gas token:

```
L1Block
The L1Block contract is upgraded with a permissioned function setGasPayingToken that is used to set information about the gas paying token. The DEPOSITOR_ACCOUNT MUST be the only account that can call the setter function. This setter is differentiated from the setL1BlockValues functions because the gas paying token is not meant to be dynamic config whereas the arguments to setL1BlockValues are generally dynamic in nature.

Any L2 contract that wants to learn the address of the gas paying token can call the getter on the L1Block contract.
```